### PR TITLE
sql: support ALTER SOURCE...ADD SUBSOURCE

### LINKED ISSUE #15832: storage: allow adding and removing tables to/from a Postgres Source
### Feature request

Following https://github.com/MaterializeInc/materialize/issues/14016, there is a limitation where you cannot add or remove a table to/from a Postgres source without dropping and recreating the source. We probably want to allow this in `ALTER SOURCE`.

### PR DESCRIPTION
**Pull request by sploiselle**
### Motivation

This PR adds a known-desirable feature. Closes #15832

### Tips for reviewer

- Reviewing commit by commit is likely best
- There are some semantic changes to `TEXT COLUMNS` baked in here that I have ensured do not interfere with users who've done now-prohibited things previously (e.g. referring to a table that is not being added to the source). These changes _do_ break the ability to put `SHOW CREATE SOURCE` back into `CREATE SOURCE`.  However, this is a weird and fixable edge case––I do not believe we should hold back making this change or add complexity to support it in `CREATE` but not `ALTER`.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Let users add new tables to their PG sources using `ALTER SOURCE...ADD SUBSOURCE`.

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
I triggered a
* [nightly build](https://buildkite.com/materialize/nightlies/builds/2724)
* [coverage build](https://buildkite.com/materialize/coverage/builds/168)

File: src/adapter/src/coord/sequencer/inner.rs [hunk 1]
```diff
@@ -3834,6 +3877,238 @@ impl Coordinator {
...
+
+                    let (data_source, status_collection_id) = match source.data_source {
+                        // Subsources use source statuses.
+                        DataSourceDesc::Source => (DataSource::Other, source_status_collection_id),
+                        o => {
+                            unreachable!(
+                                "ALTER SOURCE...ADD SUBSOURCE only creates subsources but got {:?}",
+                                o
+                            )
+                        }
```
Comment (nrainer-materialize) [inline src/adapter/src/coord/sequencer/inner.rs:4079]: Uncovered. Do you think it makes sense to cover this code with a test? Feel free to ignore if not.

File: src/sql/src/pure.rs [hunk 2]
```diff
@@ -813,7 +626,223 @@ pub async fn purify_create_source(
...
+        }
+
+        match &desc.connection {
+            GenericSourceConnection::Postgres(pg_connection) => pg_connection.clone(),
+            _ => sql_bail!(
+                "{} is a {} source, which does not support ALTER TABLE...ADD SUBSOURCES",
+                scx.catalog.minimal_qualification(item.name()),
+                desc.connection.name()
+            ),
+        }
```
Comment (nrainer-materialize) [inline src/sql/src/pure.rs:680]: Uncovered. Does it makes sense to add a test case here?

File: src/sql/src/pure.rs [hunk 3]
```diff
@@ -813,7 +626,223 @@ pub async fn purify_create_source(
...
+
+    let mut publication_tables =
+        mz_postgres_util::publication_info(&config, &pg_source_connection.publication, None)
+            .await?;
+
+    if publication_tables.is_empty() {
+        return Err(PlanError::EmptyPublication(
+            pg_source_connection.publication.to_string(),
+        ));
+    }
```
Comment (nrainer-materialize) [inline src/sql/src/pure.rs:726]: Uncovered. Shall we add a test to cover this?

File: src/adapter/src/coord/sequencer/inner.rs [hunk 4]
```diff
@@ -3834,6 +3877,238 @@ impl Coordinator {
...
+                                _ => false,
+                            },
+                            "TEXT COLUMNS must have a sequence of unresolved item names but got {:?}",
+                            new.value
+                        );
+
+                        curr_options.push(PgConfigOption {
+                            name: PgConfigOptionName::TextColumns,
+                            value: new.value,
+                        })
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/adapter/src/coord/sequencer/inner.rs:3985]: Great catch--this should be done now.

File: src/sql/src/pure.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/pure.rs:680]: Added test

File: src/sql/src/pure.rs (hunk 3 above)
Comment (sploiselle) [reply src/sql/src/pure.rs:726]: Added test.

File: src/adapter/src/coord/sequencer/inner.rs (hunk 1 above)
Comment (sploiselle) [reply src/adapter/src/coord/sequencer/inner.rs:4076]: The top line (4073 in the diff this comment was on) should be covered, but those below it should not be.

File: src/adapter/src/coord/sequencer/inner.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrainer-materialize) [reply src/adapter/src/coord/sequencer/inner.rs:3985]: Great, thanks! 👍

File: src/sql/src/pure.rs [hunk 5]
```diff
@@ -813,7 +626,223 @@ pub async fn purify_create_source(
 
     purify_source_format(&*catalog, format, connection, envelope, &connection_context).await?;
 
-    Ok((subsources, stmt))
+    Ok((subsources, Statement::CreateSource(stmt)))
+}
+
+async fn purify_alter_source(
```
Comment (jkosh44) [inline src/sql/src/pure.rs:632]: Some doc-comments would probably help understand what the return type is.

File: src/sql/src/pure.rs [hunk 6]
```diff
@@ -813,7 +626,223 @@ pub async fn purify_create_source(
...
+            GenericSourceConnection::Postgres(pg_connection) => pg_connection.clone(),
+            _ => sql_bail!(
+                "{} is a {} source, which does not support ALTER TABLE...ADD SUBSOURCES",
+                scx.catalog.minimal_qualification(item.name()),
+                desc.connection.name()
+            ),
+        }
+    };
+
+    // If there's no further work to do here, early return.
```
Comment (jkosh44) [inline src/sql/src/pure.rs:683]: Is this comment accurate? It doesn't seem like there's an early return here.

File: src/adapter/src/coord/sequencer/inner.rs [hunk 7]
```diff
@@ -3822,6 +3871,241 @@ impl Coordinator {
...
+                const ALTER_SOURCE: &str = "ALTER SOURCE...ADD SUBSOURCES";
+
+                // Resolve items in statement
+                let (mut create_source_stmt, resolved_ids) =
+                    create_sql_to_stmt_deps(self, ALTER_SOURCE, cur_entry.create_sql())?;
+
+                // We are doing a lot of unwrapping, so just make an error to reference; all of
+                // these invariants are guaranteed to be true because of how we plan subsources.
+                let purification_err =
+                    || AdapterError::internal(ALTER_SOURCE, "error in subsource purification");
```
Comment (jkosh44) [inline src/adapter/src/coord/sequencer/inner.rs:3888]: Isn't purification technically over at this point?

File: src/sql-parser/src/ast/defs/statement.rs [hunk 8]
```diff
@@ -1624,10 +1624,48 @@ impl<T: AstInfo> AstDisplay for AlterSinkStatement<T> {
...
+}
+impl_display_t!(AlterSourceAddSubsourceOption);
+
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum AlterSourceAction<T: AstInfo> {
     SetOptions(Vec<CreateSourceOption<T>>),
     ResetOptions(Vec<CreateSourceOptionName>),
+    AddSubsources {
+        subsources: Vec<CreateSourceSubsource<T>>,
+        details: Option<WithOptionValue<T>>,
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql-parser/src/ast/defs/statement.rs:1666]: It's always parsed as none but is always filled out by purification. This is just the most convenient place to store this that lets us update the existing source's details option.

**Review by jkosh44 — APPROVED** [review]

File: src/sql-parser/src/parser.rs [hunk 9]
```diff
@@ -4120,9 +4120,46 @@ impl<'a> Parser<'a> {
...
+                    // TODO: Add IF NOT EXISTS?
+                    let subsources = self
+                        .parse_comma_separated(Parser::parse_subsource_references)
+                        .map_parser_err(StatementKind::AlterSource)?;
+
+                    let options = if self.parse_keyword(WITH) {
+                        self.expect_token(&Token::LParen)
+                            .map_parser_err(StatementKind::AlterSource)?;
+                        let o = if matches!(self.peek_token(), Some(Token::RParen)) {
+                            vec![]
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql-parser/src/parser.rs]: Oh I stole this line from `parse_create_index`; will adjust this code.

File: src/sql-parser/tests/testdata/ddl
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql-parser/tests/testdata/ddl:1]: Done

File: src/sql/src/pure/postgres.rs [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+            .entry(database.to_string())
+            .or_insert(table);
+    }
+
+    Ok(ErsatzCatalog(tables_by_name))
+}
+
+pub(super) async fn validate_requested_subsources(
+    config: &Config,
+    r: &[(UnresolvedItemName, UnresolvedItemName, &PostgresTableDesc)],
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql/src/pure/postgres.rs]: Added descriptive name

File: src/sql/src/pure.rs (hunk 5 above)
Comment (sploiselle) [reply src/sql/src/pure.rs:638]: This had as much documentation as `purify_create_source` :D Added something

File: src/sql/src/pure.rs (hunk 6 above)
Comment (sploiselle) [reply src/sql/src/pure.rs:690]: Code drifted and this comment didn't follow––fixed

**Comment by sploiselle** [comment]
I preemptively rebased this over #20329 which is set to merge into `main` and which is required to ensure that this code doesn't panic when snapshotting.

**Pull request merged**